src/client/client.c

```diff
@@ -161,6 +162,129 @@ void unit_list_unref(UnitList *unit_list) {
         free(unit_list);
 }
 
+static sd_bus_message *wait_for_job(Client *client, char *object_path) {
+        int r = 0;
+
+        assert(client->waiting_for_job == NULL);
+        client->waiting_for_job = object_path;
+
+        for (;;) {
+                /* Process requests */
+                r = sd_bus_process(client->api_bus, NULL);
+                if (r < 0) {
+                        fprintf(stderr, "Failed to process bus: %s\n", strerror(-r));
+                        return NULL;
+                }
+
+                /* Did we get the result? */
+                if (client->waiting_for_job_result != NULL) {
+                        return steal_pointer(&client->waiting_for_job_result);
+                }
+
+                if (r > 0) { /* request processed, try to process another one */
+                        continue;
+                }
+
+                /* Wait for the next request to process */
+                r = sd_bus_wait(client->api_bus, (uint64_t) -1);
+                if (r < 0) {
+                        fprintf(stderr, "Failed to wait on bus: %s\n", strerror(-r));
+                        return NULL;
+                }
+        }
+}
+
+static int match_job_removed_signal(sd_bus_message *m, void *userdata, UNUSED sd_bus_error *error) {
+        const char *job_path = NULL, *result = NULL, *node = NULL, *unit = NULL;
+        uint32_t id = 0;
+        int r = 0;
+        Client *client = (Client *) userdata;
+
+        if (client->waiting_for_job == NULL) {
+                return 0;
+        }
+
+        r = sd_bus_message_read(m, "uosss", &id, &job_path, &node, &unit, &result);
+        if (r < 0) {
+                fprintf(stderr, "Can't parse job result\n");
+                return 0;
+        }
+
+        (void) sd_bus_message_rewind(m, true);
+
+        if (strcmp(client->waiting_for_job, job_path) == 0) {
+                client->waiting_for_job_result = sd_bus_message_ref(m);
+                return 1;
+        }
+
+        return 0;
+}
+
+int method_lifecycle_action_on(Client *client, char *target, char *unit, char *method) {
+        _cleanup_sd_bus_error_ sd_bus_error error = SD_BUS_ERROR_NULL;
+        _cleanup_sd_bus_message_ sd_bus_message *message = NULL;
+        _cleanup_sd_bus_message_ sd_bus_message *job_result = NULL;
+        char *job_path = NULL, *result = NULL, *node = NULL;
+        uint32_t id = 0;
+        int r = 0;
+
+        r = assemble_object_path_string(NODE_OBJECT_PATH_PREFIX, target, &client->object_path);
+        if (r < 0) {
+                return r;
+        }
+
+        r = sd_bus_match_signal(
+                        client->api_bus,
+                        NULL,
+                        HIRTE_INTERFACE_BASE_NAME,
+                        HIRTE_MANAGER_OBJECT_PATH,
+                        MANAGER_INTERFACE,
+                        "JobRemoved",
+                        match_job_removed_signal,
+                        client);
```

Comment:
Instead of passing the whole `client`, I think just passing a `sd_bus_message` (declared at the top of the `method_lifecycle_action_on`) would be sufficient. The same message variable would be passed to `wait_for_job` (instead of the client) where the message is read and the `job_path` is being checked. 
This way, `client` does not need the new member variables.

Reply:
Also, passing a pointer to client object and passing a pointer to message object is the same performance-wise. 
If we would write this in an OO language, it would be client.wait_for_job()

Reply:
We need `waiting_for_job` to know what signal to match, because it will catch all `JobRemoved` signals, but we need only the one that is related to the specific job, and this is what `waiting_for_job` contains.

Reply:
I mean, I'd prefer to have a dedicated struct for passing around this information, e.g. :
```C
struct PendingJob{
   char *job_name;
   sd_bus_message *job_result;
}
```
If you don't like this approach, can we rename the variables to sth like`pending_job_name` and `pending_job_result`? `waiting_for_*` sounds more like a function call.

Reply:
I vote for the second option. Yes, these names are clearer. Changed that